PKCS11 Wrapper

glatigny opened the pull request:
Continuation of #8 

* Split the wrapper and the PKCS11 Client
* Remove the Encapsulation in Encrypt/Decrypt
* Use "ReadBaoVariable" and update Env variables
* Update parameters (according to [developer.hashicorp.com/vault/docs/configuration/seal/pkcs11](https://developer.hashicorp.com/vault/docs/configuration/seal/pkcs11#pkcs11-parameters))
* Support of Token Label parameter
* Add tests for options
* Apply @cipherboy comments from PR #8 

Tested with SoftHSM2 (AES_CBC_PAD & RSA_PKCS keys)

cipherboy commented on wrappers/pkcs11/options.go:
```diff
@@ -0,0 +1,167 @@
+// Copyright (c) HashiCorp, Inc.
```
@glatigny Regarding copyright statements, if this is net-original code, I'd suggest a header of:

```go
// Copyright (c) 2024 OpenBao a Series of LF Projects, LLC
```

Many of these options &c are similar to existing patterns, so I'm happy with a mix of both headers as is likely the case.

cipherboy commented on wrappers/pkcs11/pkcs11_client.go:
```diff
@@ -0,0 +1,553 @@
+	case api.ReadBaoVariable(EnvHsmWrapperKeyId) != "" && !opts.Options.WithDisallowEnvVars:
+		keyId = api.ReadBaoVariable(EnvHsmWrapperKeyId)
+	case opts.withKeyId != "":
+		keyId = opts.withKeyId
```
I think this comment might still apply? https://github.com/openbao/go-kms-wrapping/pull/8/files/45ad5ab66ef39ae498a8a736b1a190d34ca1c397#r1718338953

glatigny replied:
I added some code to remove the prefix on variable `keyId` after the switch.

cipherboy commented on wrappers/pkcs11/pkcs11_client.go:
```diff
@@ -0,0 +1,553 @@
+	c.client.CloseSession(session)
+}
+
+//
```
nit: empty comment?

cipherboy commented on wrappers/pkcs11/pkcs11_client.go:
```diff
@@ -0,0 +1,553 @@
+	if err := c.client.FindObjectsInit(session, template); err != nil {
+		return nil, fmt.Errorf("failed to pkcs11 FindObjectsInit: %s", err)
+	}
+	obj, _, err := c.client.FindObjects(session, 2)
```
How do we want to handle multiple results? I think right now, given the above comment about uniqueness, we're at the mercy of the library as to what order it returns results in, which might be random.

Thoughts?

glatigny replied:
There is a part in `Encrypt` and `Decrypt` that I'll move into `FindKey`: (and I will change the text too).
```
	if len(obj) != 1 {
		return nil, nil, fmt.Errorf("expected 1 object, got %d", len(obj))
	}
```
Since we're asking for 2 objects max, we can detect if there is more than 1 and then return an error.

cipherboy commented on wrappers/pkcs11/pkcs11_client.go:
```diff
@@ -0,0 +1,553 @@
+		if c.mechanism != 0 {
+			mechanism = c.mechanism
+		} else {
+			mechanism = pkcs11.CKM_AES_GCM
```
Thanks, @glatigny!!

cipherboy reviewed (commented):
Sorry about the delay! This is looking fairly good, thank you! I've left a few comments here and there.

cipherboy commented on wrappers/pkcs11/pkcs11_client.go:
```diff
@@ -0,0 +1,553 @@
+		return 0, fmt.Errorf("PKCS11 not initialized")
+	}
+
+	if c.slot == 0 {
```
@glatigny Ah, another bug. SoftHSM's default initialization suggests slot 0, so I'd suggest trying it first rather than immediately going to `GetSlotForLabel()`:

```
$ softhsm2-util --init-token --slot 0 --label "My token 1"
```

per https://wiki.opendnssec.org/softhsm2/

cipherboy replied:
My bad, I forgot SoftHSM assigns random IDs in that case:

```
Available slots:
Slot 1513794117
    Slot info:
        Description:      SoftHSM slot ID 0x5a3aaa45                                      
        Manufacturer ID:  SoftHSM project                 
        Hardware version: 2.6
        Firmware version: 2.6
        Token present:    yes
    Token info:
        Manufacturer ID:  SoftHSM project                 
        Model:            SoftHSM v2      
        Hardware version: 2.6
        Firmware version: 2.6
        Serial number:    f521fa3b5a3aaa45
        Initialized:      yes
        User PIN init.:   yes
        Label:            My token 1                      
Slot 1
    Slot info:
        Description:      SoftHSM slot ID 0x1                                             
        Manufacturer ID:  SoftHSM project                 
        Hardware version: 2.6
        Firmware version: 2.6
        Token present:    yes
    Token info:
        Manufacturer ID:  SoftHSM project                 
        Model:            SoftHSM v2      
        Hardware version: 2.6
        Firmware version: 2.6
        Serial number:                    
        Initialized:      no
        User PIN init.:   no
        Label:                                            
```

cipherboy commented on wrappers/pkcs11/pkcs11_client.go:
```diff
@@ -0,0 +1,553 @@
+		return nil, nil, err
+	}
+
+	var iv []byte
```
@glatigny I think you need to correctly handle mechanism parameters here. See https://pkg.go.dev/github.com/miekg/pkcs11#NewGCMParams or https://pkg.go.dev/github.com/miekg/pkcs11#NewOAEPParams.

Otherwise, this fails for me:

```
    pkcs11_acc_test.go:32: err : failed to pkcs11 EncryptInit: pkcs11: 0x7: CKR_ARGUMENTS_BAD
```

cipherboy replied:
I ended up with a Dockerfile:

```
FROM registry.fedoraproject.org/fedora:latest

COPY . /go-kms-wrapping


# pkcs11-tool --module "$BAO_HSM_LIB" --token-label "$BAO_HSM_TOKEN_LABEL" --so-pin 1234 --pin 4321 --keygen --key-type aes:32 --label "$BAO_HSM_KEY_LABEL" && \

# Set up SoftHSM, generate a key, and run tests.
RUN true && \
    dnf update --refresh -y && \
    dnf install -y softhsm opensc golang && \
    mkdir /softhsm && \
    export SOFTHSM2_CONF=/go-kms-wrapping/wrappers/pkcs11/softhsm2.conf && \
    softhsm2-util --init-token --slot 0 --label "OpenBao g-k-w" --so-pin 1234 --pin 4321 && \
    softhsm2-util --show-slots && \
    export BAO_HSM_TOKEN_LABEL="OpenBao g-k-w" && \
    export BAO_HSM_PIN=4321 && \
    export BAO_HSM_LIB=/usr/lib64/softhsm/libsofthsm.so && \
    export BAO_HSM_KEY_LABEL=bao-root-key && \
    pkcs11-tool --module "$BAO_HSM_LIB" --token-label "$BAO_HSM_TOKEN_LABEL" --so-pin 1234 --pin 4321 --keypairgen --key-type rsa:4096 --label "$BAO_HSM_KEY_LABEL" && \
    pkcs11-tool --module "$BAO_HSM_LIB" --token-label "$BAO_HSM_TOKEN_LABEL" --so-pin 1234 --pin 4321 --show-info && \
    cd /go-kms-wrapping/wrappers/pkcs11 && \
    export VAULT_ACC=true && \
    go test -v ./... && \
    true
```

with config (`wrappers/pkcs11/softhsm2.conf`):

```
directories.tokendir = /softhsm
objectstore.backend = file
log.level = DEBUG
```

and run via:

```
$ podman build -f wrappers/pkcs11/Dockerfile .
```

dangerousd75 commented:
I am prepared to test this with Thales USB HSM, NitroKey HSM and Yubico 2 HSM in December.
My Entrust Edge will only arrive in February.

glatigny commented on wrappers/pkcs11/pkcs11_client.go:
```diff
@@ -0,0 +1,553 @@
+	wrapConfig := new(wrapping.WrapperConfig)
+	wrapConfig.Metadata = make(map[string]string)
+	wrapConfig.Metadata["lib"] = lib
+	wrapConfig.Metadata["key_label"] = keyLabel
```
Indeed the Key Label (`CKA_LABEL`) and Key ID (`CKA_ID`) are not unique ; the user is free to choose them (and edit them).
The Unique ID is unique by definition and generated by the Token/HSM itself ; which means that if you import your secret the Unique ID will change.
That's why, I'm not a fan of using the Unique ID in the configuration.

glatigny replied:
Label + ID + KeyType muse be unique yes.
Now, if the ID is not provide during encryption and we found one key for that label, we can retrieve the ID to store it in the metadata, so during the decryption, we will be able to find the right key even if another one has been created with the same label.
If there is more than one key, the fonction FindKey will return an error.

glatigny commented on wrappers/pkcs11/pkcs11.go:
```diff
@@ -0,0 +1,123 @@
+	k.keyId = client.GetCurrentKey().String()
+
+	// Send a value to test the wrapper and to set the current key id
+	if _, err := k.Encrypt(context.Background(), []byte("a")); err != nil {
```
I think I'll remove the encryption test here. The PKCS11 client already test the session.
Regarding the mechanisms, of course AEAD are better.

Something currently wrong in the implementation is the parameter `hmac_mechanism` is not handled and should be required for CBC and CBC PAD.
At first, I was thinking that SoftHSM2 did not support AES GCM, but it's just the documentation which is not up-to-date.
There is still the issue with YubiHSM2 but yes, it might be interesting to review the list of supported mechanisms !

cipherboy replied:
I'm not inclined to support hmac parameter and just require an AEAD mech.

glatigny replied:
Yes, of course.
It just like you wrote 3 days ago: to support encrypt-then-mac it requires to add the HMAC mechanism parameter.

glatigny commented on wrappers/pkcs11/pkcs11_client.go:
```diff
@@ -0,0 +1,553 @@
+	switch mech {
+	case pkcs11.CKM_AES_GCM:
+		return true, 16
+	case pkcs11.CKM_AES_CBC_PAD:
```
Bad news... It is not possible to add `CCM` support for the moment.

The PKCS11 lib doesn't have the function to create the CCM parameters:
https://github.com/miekg/pkcs11/blob/master/params.go
but the struct is available in the headers:
https://github.com/miekg/pkcs11/blob/master/pkcs11t.h#L1888
So it needs a small PR to add the wrapping struct and functions.

cipherboy replied:
@glatigny Hmm.... Sadly it kinda looks like miekg/pkcs11 is getting less active. Things seemed to stall after attempting the PKCS#11 v3.0 port: https://github.com/miekg/pkcs11/pull/148 ... 

I'm not aware of any good alternatives to this library, so perhaps we should start trying to get involved upstream?

cipherboy commented on wrappers/pkcs11/pkcs11_client.go:
```diff
@@ -0,0 +1,632 @@
+	if err != nil {
+		return nil, nil, err
+	}
+	if err = c.client.EncryptInit(session, []*pkcs11.Mechanism{params}, key); err != nil {
```
@glatigny Hmm, this currently panics with AES-GCM and the Dockerfile above... 

```
/usr/include/c++/14/bits/stl_vector.h:1127: std::vector<_Tp, _Alloc>::reference std::vector<_Tp, _Alloc>::operator[](size_type) [with _Tp = unsigned char; _Alloc = SecureAllocator<unsigned char>; reference = unsigned char&; size_type = long unsigned int]: Assertion '__n < this->size()' failed.
SIGABRT: abort
PC=0x7f83f2642c94 m=0 sigcode=18446744073709551610
signal arrived during cgo execution

goroutine 8 gp=0xc00018e8c0 m=0 mp=0xae6780 [syscall]:
runtime.cgocall(0x734cf0, 0xc000086c38)
	/usr/lib/golang/src/runtime/cgocall.go:167 +0x4b fp=0xc000086c10 sp=0xc000086bd8 pc=0x46bb2b
github.com/miekg/pkcs11._Cfunc_EncryptInit(0x3ef3bf50, 0x2, 0xc0000123f0, 0x3)
	_cgo_gotypes.go:873 +0x4c fp=0xc000086c38 sp=0xc000086c10 pc=0x719ccc
github.com/miekg/pkcs11.(*Ctx).EncryptInit.func1(0xc000086dd0?, 0x2, 0xc0000123f0, 0x3)
	/root/go/pkg/mod/github.com/miekg/pkcs11@v1.1.2-0.20231115102856-9078ad6b9d4b/pkcs11.go:1155 +0x7d fp=0xc000086c78 sp=0xc000086c38 pc=0x71d77d
github.com/miekg/pkcs11.(*Ctx).EncryptInit(0xc000072220, 0x2, {0xc000086dd0?, 0x7f83f2596748?, 0x0?}, 0x3)
	/root/go/pkg/mod/github.com/miekg/pkcs11@v1.1.2-0.20231115102856-9078ad6b9d4b/pkcs11.go:1155 +0x99 fp=0xc000086ce8 sp=0xc000086c78 pc=0x71d659
github.com/openbao/go-kms-wrapping/wrappers/pkcs11/v2.(*Pkcs11Client).Encrypt(0xc0000e5780, {0xc0001cb978, 0x3, 0x3})
	/go-kms-wrapping/wrappers/pkcs11/pkcs11_client.go:249 +0x1a5 fp=0xc000086df8 sp=0xc000086ce8 pc=0x72f205
github.com/openbao/go-kms-wrapping/wrappers/pkcs11/v2.(*Wrapper).Encrypt(0xc000086f08, {0x762220?, 0x87d070?}, {0xc0001cb978?, 0x50a200?, 0xa2ccd8?}, {0xaaff60?, 0xc0001cb8a8?, 0x50a22d?})
	/go-kms-wrapping/wrappers/pkcs11/pkcs11.go:76 +0x43 fp=0xc000086e90 sp=0xc000086df8 pc=0x72dc63
github.com/openbao/go-kms-wrapping/wrappers/pkcs11/v2.TestAccPkcs11Wrapper_Lifecycle(0xc0000dcea0)
	/go-kms-wrapping/wrappers/pkcs11/pkcs11_acc_test.go:36 +0x155 fp=0xc000086f70 sp=0xc000086e90 pc=0x7318f5
testing.tRunner(0xc0000dcea0, 0x8197c0)
	/usr/lib/golang/src/testing/testing.go:1690 +0xf4 fp=0xc000086fc0 sp=0xc000086f70 pc=0x50a2f4
```

glatigny replied:
Okay, I will rollback. I wanted to avoid returning a `[]*pkcs11.Mechanism`., but that's not a safe choice..
That's strange that it didn't failed on my laptop.

cipherboy replied:
Sorry, see the comment below, this was crashing because my SoftHSM on Fedora 40 didn't like the 16-byte prefix for GCM. It didn't have to do with this.

glatigny replied:
According to Oasis Documentation, we can safely use a 12 bytes length for the nonce:
https://docs.oasis-open.org/pkcs11/pkcs11-curr/v3.0/pkcs11-curr-v3.0.html#_Toc228894714

Since GCM has some specific behavior (free params, IV can be added by the HSM...) ; I'm performing some refactoring.
I'll see if I can keep some common code but it's possible that I'll split AES and RSA encrypt/decrypt in dedicated functions.

cipherboy commented on wrappers/pkcs11/pkcs11_client.go:
```diff
@@ -0,0 +1,632 @@
+	}
+	switch mechanism {
+	case pkcs11.CKM_AES_GCM:
+		return mechanism, 16, nil
```
@glatigny Now _that_ is unintuitive. The panic occurs in SoftHSM because we passed 16 bytes instead of 12 here.

cipherboy commented on wrappers/pkcs11/pkcs11_client.go:
```diff
@@ -0,0 +1,632 @@
+	var params interface{}
+	switch mechanism {
+	case pkcs11.CKM_AES_GCM:
+		params = pkcs11.NewGCMParams(prefix, nil, 128)
```
Hmm, check out this note: https://pkg.go.dev/github.com/miekg/pkcs11#NewGCMParams

> Note that some HSMs, like CloudHSM, will ignore the IV you pass in and write their own. As a result, to support all libraries, memory is not freed automatically, so that after the EncryptInit/Encrypt operation the HSM's IV can be read back out. It is up to the caller to ensure that Free() is called on the GCMParams object at an appropriate time, which is after Encrypt/Decrypt. As an example: 

.... so we likely want to make sure we save `params` here and update our copy of `prefix` from this value before doing the append in `Encrypt(...)`.

glatigny commented:
Hey @cipherboy
The last commit changes introduce some refactoring.

I wanted to make the constants a bit cleaner so I moved them into a dedicated section.
At the same time, I changed the IV/Nonce length for AES GCM.

While spliting the Encrypt/Decrypt functions for AES GCM and RSA OAEP, I saw how the IV could be handle by CloudHSM (and maybe others) and also saw that BlobInfo can embed the IV.

I also remove the usage of `uuid` in order to use the random generator via PKCS#11.
Valut has the "Entropy augmentation seal" feature, the `GenerateRandom` function could be a start for it.

cipherboy commented:
@glatigny Hmm, what version of SoftHSM are you testing against? I've been pulling `softhsm                    x86_64 2.6.1-9.fc40` [in the container](https://github.com/openbao/go-kms-wrapping/pull/14#discussion_r1857158113) but both AES and RSA fail:

<details>

<summary> AES failure </summary>

```
>>> Creating user 'systemd-timesync' (systemd Time Synchronization) with UID 996
>>> 
Complete!
The token has been initialized and is reassigned to slot 1000552664
Available slots:
Slot 1000552664
    Slot info:
        Description:      SoftHSM slot ID 0x3ba338d8                                      
        Manufacturer ID:  SoftHSM project                 
        Hardware version: 2.6
        Firmware version: 2.6
        Token present:    yes
    Token info:
        Manufacturer ID:  SoftHSM project                 
        Model:            SoftHSM v2      
        Hardware version: 2.6
        Firmware version: 2.6
        Serial number:    789e33f33ba338d8
        Initialized:      yes
        User PIN init.:   yes
        Label:            OpenBao g-k-w                   
Slot 1
    Slot info:
        Description:      SoftHSM slot ID 0x1                                             
        Manufacturer ID:  SoftHSM project                 
        Hardware version: 2.6
        Firmware version: 2.6
        Token present:    yes
    Token info:
        Manufacturer ID:  SoftHSM project                 
        Model:            SoftHSM v2      
        Hardware version: 2.6
        Firmware version: 2.6
        Serial number:                    
        Initialized:      no
        User PIN init.:   no
        Label:                                            
warning: PKCS11 function C_GetAttributeValue(VALUE) failed: rv = CKR_ATTRIBUTE_SENSITIVE (0x11)

warning: PKCS11 function C_GetAttributeValue(VERIFY_RECOVER) failed: rv = CKR_ATTRIBUTE_TYPE_INVALID (0x12)

Key generated:
Secret Key Object; AES length 32
  label:      bao-root-key
  Usage:      encrypt, decrypt, sign, verify, wrap, unwrap
  Access:     never extractable, local
  uri:        pkcs11:model=SoftHSM%20v2;manufacturer=SoftHSM%20project;serial=789e33f33ba338d8;token=OpenBao%20g-k-w;object=bao-root-key;type=secret-key
Cryptoki version 2.40
Manufacturer     SoftHSM
Library          Implementation of PKCS11 (ver 2.6)
go: downloading github.com/stretchr/testify v1.8.4
go: downloading github.com/openbao/openbao/api/v2 v2.0.1
go: downloading github.com/miekg/pkcs11 v1.1.2-0.20231115102856-9078ad6b9d4b
go: downloading github.com/hashicorp/go-uuid v1.0.3
go: downloading google.golang.org/protobuf v1.33.0
go: downloading github.com/pmezard/go-difflib v1.0.0
go: downloading gopkg.in/yaml.v3 v3.0.1
go: downloading github.com/davecgh/go-spew v1.1.1
go: downloading golang.org/x/time v0.0.0-20220411224347-583f2d630306
go: downloading github.com/hashicorp/go-retryablehttp v0.7.7
go: downloading github.com/mitchellh/mapstructure v1.5.0
go: downloading github.com/hashicorp/go-cleanhttp v0.5.2
go: downloading github.com/hashicorp/go-rootcerts v1.0.2
go: downloading github.com/cenkalti/backoff/v3 v3.0.0
go: downloading github.com/hashicorp/hcl v1.0.0
go: downloading golang.org/x/net v0.26.0
go: downloading github.com/hashicorp/go-multierror v1.1.1
go: downloading github.com/go-jose/go-jose/v3 v3.0.1
go: downloading github.com/hashicorp/errwrap v1.1.0
go: downloading github.com/hashicorp/go-secure-stdlib/parseutil v0.1.6
go: downloading github.com/hashicorp/go-sockaddr v1.0.2
go: downloading github.com/hashicorp/go-secure-stdlib/strutil v0.1.2
go: downloading golang.org/x/crypto v0.24.0
go: downloading github.com/ryanuber/go-glob v1.0.0
go: downloading golang.org/x/text v0.16.0
=== RUN   Test_GetOpts
=== PAUSE Test_GetOpts
=== RUN   TestAccPkcs11Wrapper_Lifecycle
/usr/include/c++/14/bits/stl_vector.h:1127: std::vector<_Tp, _Alloc>::reference std::vector<_Tp, _Alloc>::operator[](size_type) [with _Tp = unsigned char; _Alloc = SecureAllocator<unsigned char>; reference = unsigned char&; size_type = long unsigned int]: Assertion '__n < this->size()' failed.
SIGABRT: abort
PC=0x7f3d19e700f4 m=0 sigcode=18446744073709551610
signal arrived during cgo execution

goroutine 8 gp=0xc00018e8c0 m=0 mp=0xae77a0 [syscall]:
runtime.cgocall(0x735b90, 0xc000086b20)
	/usr/lib/golang/src/runtime/cgocall.go:167 +0x4b fp=0xc000086af8 sp=0xc000086ac0 pc=0x46bc8b
github.com/miekg/pkcs11._Cfunc_EncryptInit(0x350a4f50, 0x2, 0xc0000123f0, 0x3)
	_cgo_gotypes.go:873 +0x4c fp=0xc000086b20 sp=0xc000086af8 pc=0x719d2c
github.com/miekg/pkcs11.(*Ctx).EncryptInit.func1(0xc000086c68?, 0x2, 0xc0000123f0, 0x3)
	/root/go/pkg/mod/github.com/miekg/pkcs11@v1.1.2-0.20231115102856-9078ad6b9d4b/pkcs11.go:1155 +0x7d fp=0xc000086b60 sp=0xc000086b20 pc=0x71d8fd
github.com/miekg/pkcs11.(*Ctx).EncryptInit(0xc000072220, 0x2, {0xc000086c68?, 0x0?, 0x0?}, 0x3)
	/root/go/pkg/mod/github.com/miekg/pkcs11@v1.1.2-0.20231115102856-9078ad6b9d4b/pkcs11.go:1155 +0x99 fp=0xc000086bd0 sp=0xc000086b60 pc=0x71d7d9
github.com/openbao/go-kms-wrapping/wrappers/pkcs11/v2.(*Pkcs11Client).EncryptAesGcm(0xc0000e5780, 0x2, 0x3, {{0xc000020012, 0xc}, {0x0, 0x0}}, {0xc0001cb978, 0x3, 0x3})
	/go-kms-wrapping/wrappers/pkcs11/pkcs11_client.go:291 +0x22c fp=0xc000086ce0 sp=0xc000086bd0 pc=0x72fe0c
github.com/openbao/go-kms-wrapping/wrappers/pkcs11/v2.(*Pkcs11Client).Encrypt(0xc0000e5780, {0xc0001cb978, 0x3, 0x3})
	/go-kms-wrapping/wrappers/pkcs11/pkcs11_client.go:270 +0x1ad fp=0xc000086de0 sp=0xc000086ce0 pc=0x72f6ad
github.com/openbao/go-kms-wrapping/wrappers/pkcs11/v2.(*Wrapper).Encrypt(0xc000086f08, {0x763100?, 0x87e140?}, {0xc0001cb978?, 0x50a3a0?, 0xa2e270?}, {0xab0f60?, 0xc0001cb8a8?, 0x50a3cd?})
	/go-kms-wrapping/wrappers/pkcs11/pkcs11.go:76 +0x43 fp=0xc000086e90 sp=0xc000086de0 pc=0x72e0a3
github.com/openbao/go-kms-wrapping/wrappers/pkcs11/v2.TestAccPkcs11Wrapper_Lifecycle(0xc0000dcea0)
	/go-kms-wrapping/wrappers/pkcs11/pkcs11_acc_test.go:36 +0x155 fp=0xc000086f70 sp=0xc000086e90 pc=0x732795
testing.tRunner(0xc0000dcea0, 0x81a5e8)
	/usr/lib/golang/src/testing/testing.go:1690 +0xf4 fp=0xc000086fc0 sp=0xc000086f70 pc=0x50a494
testing.(*T).Run.gowrap1()
	/usr/lib/golang/src/testing/testing.go:1743 +0x25 fp=0xc000086fe0 sp=0xc000086fc0 pc=0x50b485
runtime.goexit({})
	/usr/lib/golang/src/runtime/asm_amd64.s:1700 +0x1 fp=0xc000086fe8 sp=0xc000086fe0 pc=0x479e21
created by testing.(*T).Run in goroutine 1
	/usr/lib/golang/src/testing/testing.go:1743 +0x390

goroutine 1 gp=0xc0000061c0 m=nil [chan receive]:
runtime.gopark(0xb9790?, 0x7f3d19dc31a8?, 0x18?, 0x0?, 0x78b1e0?)
	/usr/lib/golang/src/runtime/proc.go:424 +0xce fp=0xc0001879c8 sp=0xc0001879a8 pc=0x4721ce
runtime.chanrecv(0xc0000ab260, 0xc000187aaf, 0x1)
	/usr/lib/golang/src/runtime/chan.go:639 +0x41c fp=0xc000187a40 sp=0xc0001879c8 pc=0x40855c
runtime.chanrecv1(0xae5ae0?, 0x763100?)
	/usr/lib/golang/src/runtime/chan.go:489 +0x12 fp=0xc000187a68 sp=0xc000187a40 pc=0x408132
testing.(*T).Run(0xc0000dcb60, {0x7f18e8?, 0x0?}, 0x81a5e8)
	/usr/lib/golang/src/testing/testing.go:1751 +0x3ab fp=0xc000187b28 sp=0xc000187a68 pc=0x50b32b
testing.runTests.func1(0xc0000dcb60)
	/usr/lib/golang/src/testing/testing.go:2168 +0x37 fp=0xc000187b68 sp=0xc000187b28 pc=0x50d5f7
testing.tRunner(0xc0000dcb60, 0xc000187c70)
	/usr/lib/golang/src/testing/testing.go:1690 +0xf4 fp=0xc000187bb8 sp=0xc000187b68 pc=0x50a494
testing.runTests(0xc000012318, {0xa9caa0, 0x2, 0x2}, {0x470050?, 0x46fcba?, 0xae6480?})
	/usr/lib/golang/src/testing/testing.go:2166 +0x43d fp=0xc000187ca0 sp=0xc000187bb8 pc=0x50d4dd
testing.(*M).Run(0xc0001e00a0)
	/usr/lib/golang/src/testing/testing.go:2034 +0x64a fp=0xc000187ed0 sp=0xc000187ca0 pc=0x50bf0a
main.main()
	_testmain.go:47 +0x9b fp=0xc000187f50 sp=0xc000187ed0 pc=0x73427b
runtime.main()
	/usr/lib/golang/src/runtime/proc.go:272 +0x28b fp=0xc000187fe0 sp=0xc000187f50 pc=0x43c2cb
runtime.goexit({})
	/usr/lib/golang/src/runtime/asm_amd64.s:1700 +0x1 fp=0xc000187fe8 sp=0xc000187fe0 pc=0x479e21

goroutine 2 gp=0xc000006c40 m=nil [force gc (idle)]:
runtime.gopark(0x0?, 0x0?, 0x0?, 0x0?, 0x0?)
	/usr/lib/golang/src/runtime/proc.go:424 +0xce fp=0xc00006efa8 sp=0xc00006ef88 pc=0x4721ce
runtime.goparkunlock(...)
	/usr/lib/golang/src/runtime/proc.go:430
runtime.forcegchelper()
	/usr/lib/golang/src/runtime/proc.go:337 +0xb3 fp=0xc00006efe0 sp=0xc00006efa8 pc=0x43c613
runtime.goexit({})
	/usr/lib/golang/src/runtime/asm_amd64.s:1700 +0x1 fp=0xc00006efe8 sp=0xc00006efe0 pc=0x479e21
created by runtime.init.7 in goroutine 1
	/usr/lib/golang/src/runtime/proc.go:325 +0x1a

goroutine 3 gp=0xc000007180 m=nil [GC sweep wait]:
runtime.gopark(0x0?, 0x0?, 0x0?, 0x0?, 0x0?)
	/usr/lib/golang/src/runtime/proc.go:424 +0xce fp=0xc00006f780 sp=0xc00006f760 pc=0x4721ce
runtime.goparkunlock(...)
	/usr/lib/golang/src/runtime/proc.go:430
runtime.bgsweep(0xc00009c000)
	/usr/lib/golang/src/runtime/mgcsweep.go:277 +0x94 fp=0xc00006f7c8 sp=0xc00006f780 pc=0x4242b4
runtime.gcenable.gowrap1()
	/usr/lib/golang/src/runtime/mgc.go:204 +0x25 fp=0xc00006f7e0 sp=0xc00006f7c8 pc=0x4189c5
runtime.goexit({})
	/usr/lib/golang/src/runtime/asm_amd64.s:1700 +0x1 fp=0xc00006f7e8 sp=0xc00006f7e0 pc=0x479e21
created by runtime.gcenable in goroutine 1
	/usr/lib/golang/src/runtime/mgc.go:204 +0x66

goroutine 4 gp=0xc000007340 m=nil [GC scavenge wait]:
runtime.gopark(0xc00009c000?, 0x87bfd0?, 0x1?, 0x0?, 0xc000007340?)
	/usr/lib/golang/src/runtime/proc.go:424 +0xce fp=0xc00006ff78 sp=0xc00006ff58 pc=0x4721ce
runtime.goparkunlock(...)
	/usr/lib/golang/src/runtime/proc.go:430
runtime.(*scavengerState).park(0xae6500)
	/usr/lib/golang/src/runtime/mgcscavenge.go:425 +0x49 fp=0xc00006ffa8 sp=0xc00006ff78 pc=0x421ce9
runtime.bgscavenge(0xc00009c000)
	/usr/lib/golang/src/runtime/mgcscavenge.go:653 +0x3c fp=0xc00006ffc8 sp=0xc00006ffa8 pc=0x42225c
runtime.gcenable.gowrap2()
	/usr/lib/golang/src/runtime/mgc.go:205 +0x25 fp=0xc00006ffe0 sp=0xc00006ffc8 pc=0x418965
runtime.goexit({})
	/usr/lib/golang/src/runtime/asm_amd64.s:1700 +0x1 fp=0xc00006ffe8 sp=0xc00006ffe0 pc=0x479e21
created by runtime.gcenable in goroutine 1
	/usr/lib/golang/src/runtime/mgc.go:205 +0xa5

goroutine 5 gp=0xc000007c00 m=nil [finalizer wait]:
runtime.gopark(0xc00006e648?, 0x40ef05?, 0xb0?, 0x1?, 0xc0000061c0?)
	/usr/lib/golang/src/runtime/proc.go:424 +0xce fp=0xc00006e620 sp=0xc00006e600 pc=0x4721ce
runtime.runfinq()
	/usr/lib/golang/src/runtime/mfinal.go:193 +0x107 fp=0xc00006e7e0 sp=0xc00006e620 pc=0x417a47
runtime.goexit({})
	/usr/lib/golang/src/runtime/asm_amd64.s:1700 +0x1 fp=0xc00006e7e8 sp=0xc00006e7e0 pc=0x479e21
created by runtime.createfing in goroutine 1
	/usr/lib/golang/src/runtime/mfinal.go:163 +0x3d

goroutine 6 gp=0xc00018e540 m=nil [chan receive]:
runtime.gopark(0x0?, 0x0?, 0x0?, 0x0?, 0x0?)
	/usr/lib/golang/src/runtime/proc.go:424 +0xce fp=0xc000070718 sp=0xc0000706f8 pc=0x4721ce
runtime.chanrecv(0xc0000aa4d0, 0x0, 0x1)
	/usr/lib/golang/src/runtime/chan.go:639 +0x41c fp=0xc000070790 sp=0xc000070718 pc=0x40855c
runtime.chanrecv1(0x0?, 0x0?)
	/usr/lib/golang/src/runtime/chan.go:489 +0x12 fp=0xc0000707b8 sp=0xc000070790 pc=0x408132
runtime.unique_runtime_registerUniqueMapCleanup.func1(...)
	/usr/lib/golang/src/runtime/mgc.go:1781
runtime.unique_runtime_registerUniqueMapCleanup.gowrap1()
	/usr/lib/golang/src/runtime/mgc.go:1784 +0x2f fp=0xc0000707e0 sp=0xc0000707b8 pc=0x41b9ef
runtime.goexit({})
	/usr/lib/golang/src/runtime/asm_amd64.s:1700 +0x1 fp=0xc0000707e8 sp=0xc0000707e0 pc=0x479e21
created by unique.runtime_registerUniqueMapCleanup in goroutine 1
	/usr/lib/golang/src/runtime/mgc.go:1779 +0x96

goroutine 7 gp=0xc00018e700 m=nil [chan receive]:
runtime.gopark(0xc0000b9748?, 0xc0000a0000?, 0x0?, 0x0?, 0xc000085e58?)
	/usr/lib/golang/src/runtime/proc.go:424 +0xce fp=0xc000085e08 sp=0xc000085de8 pc=0x4721ce
runtime.chanrecv(0xc0000ab0a0, 0x0, 0x1)
	/usr/lib/golang/src/runtime/chan.go:639 +0x41c fp=0xc000085e80 sp=0xc000085e08 pc=0x40855c
runtime.chanrecv1(0xae5ae0?, 0x763100?)
	/usr/lib/golang/src/runtime/chan.go:489 +0x12 fp=0xc000085ea8 sp=0xc000085e80 pc=0x408132
testing.(*T).Parallel(0xc0000dcd00)
	/usr/lib/golang/src/testing/testing.go:1484 +0x215 fp=0xc000085f40 sp=0xc000085ea8 pc=0x50a1f5
github.com/openbao/go-kms-wrapping/wrappers/pkcs11/v2.Test_GetOpts(0xc0000dcd00)
	/go-kms-wrapping/wrappers/pkcs11/options_test.go:16 +0x1c fp=0xc000085f70 sp=0xc000085f40 pc=0x73255c
testing.tRunner(0xc0000dcd00, 0x81a5f0)
	/usr/lib/golang/src/testing/testing.go:1690 +0xf4 fp=0xc000085fc0 sp=0xc000085f70 pc=0x50a494
testing.(*T).Run.gowrap1()
	/usr/lib/golang/src/testing/testing.go:1743 +0x25 fp=0xc000085fe0 sp=0xc000085fc0 pc=0x50b485
runtime.goexit({})
	/usr/lib/golang/src/runtime/asm_amd64.s:1700 +0x1 fp=0xc000085fe8 sp=0xc000085fe0 pc=0x479e21
created by testing.(*T).Run in goroutine 1
	/usr/lib/golang/src/testing/testing.go:1743 +0x390

rax    0x0
rbx    0x167f
rcx    0x7f3d19e700f4
rdx    0x6
rdi    0x167f
rsi    0x167f
rbp    0x7ffee090e380
rsp    0x7ffee090e340
r8     0x1
r9     0xffffffff
r10    0x8
r11    0x246
r12    0x7f3d19dfa740
r13    0x350d1710
r14    0x6
r15    0x0
rip    0x7f3d19e700f4
rflags 0x246
cs     0x33
fs     0x0
gs     0x0
FAIL	github.com/openbao/go-kms-wrapping/wrappers/pkcs11/v2	0.014s
FAIL
Error: building at STEP "RUN true && 	dnf update --refresh -y && 	dnf install -y softhsm opensc golang && 	mkdir /softhsm && 	export SOFTHSM2_CONF=/go-kms-wrapping/wrappers/pkcs11/softhsm2.conf && 	softhsm2-util --init-token --slot 0 --label "OpenBao g-k-w" --so-pin 1234 --pin 4321 && 	softhsm2-util --show-slots && 	export BAO_HSM_TOKEN_LABEL="OpenBao g-k-w" && 	export BAO_HSM_PIN=4321 && 	export BAO_HSM_LIB=/usr/lib64/softhsm/libsofthsm.so && 	export BAO_HSM_KEY_LABEL=bao-root-key &&     pkcs11-tool --module "$BAO_HSM_LIB" --token-label "$BAO_HSM_TOKEN_LABEL" --so-pin 1234 --pin 4321 --keygen --key-type aes:32 --label "$BAO_HSM_KEY_LABEL" && 	pkcs11-tool --module "$BAO_HSM_LIB" --token-label "$BAO_HSM_TOKEN_LABEL" --so-pin 1234 --pin 4321 --show-info && 	cd /go-kms-wrapping/wrappers/pkcs11 && 	export VAULT_ACC=true && 	go test -v ./... && 	true": while running runtime: exit status 1
```

</details>

<details>

<summary> RSA failure </summary>

```
>>> Creating user 'systemd-resolve' (systemd Resolver) with UID 193 and GID 193.
>>> Creating group 'systemd-timesync' with GID 996.
>>> Creating user 'systemd-timesync' (systemd Time Synchronization) with UID 996
>>> 
Complete!
The token has been initialized and is reassigned to slot 44935573
Available slots:
Slot 44935573
    Slot info:
        Description:      SoftHSM slot ID 0x2ada995                                       
        Manufacturer ID:  SoftHSM project                 
        Hardware version: 2.6
        Firmware version: 2.6
        Token present:    yes
    Token info:
        Manufacturer ID:  SoftHSM project                 
        Model:            SoftHSM v2      
        Hardware version: 2.6
        Firmware version: 2.6
        Serial number:    c1dd2d1382ada995
        Initialized:      yes
        User PIN init.:   yes
        Label:            OpenBao g-k-w                   
Slot 1
    Slot info:
        Description:      SoftHSM slot ID 0x1                                             
        Manufacturer ID:  SoftHSM project                 
        Hardware version: 2.6
        Firmware version: 2.6
        Token present:    yes
    Token info:
        Manufacturer ID:  SoftHSM project                 
        Model:            SoftHSM v2      
        Hardware version: 2.6
        Firmware version: 2.6
        Serial number:                    
        Initialized:      no
        User PIN init.:   no
        Label:                                            
Key pair generated:
Private Key Object; RSA 
  label:      bao-root-key
  Usage:      decrypt, sign, signRecover, unwrap
  Access:     sensitive, always sensitive, never extractable, local
  uri:        pkcs11:model=SoftHSM%20v2;manufacturer=SoftHSM%20project;serial=c1dd2d1382ada995;token=OpenBao%20g-k-w;object=bao-root-key;type=private
Public Key Object; RSA 4096 bits
  label:      bao-root-key
  Usage:      encrypt, verify, verifyRecover, wrap
  Access:     local
  uri:        pkcs11:model=SoftHSM%20v2;manufacturer=SoftHSM%20project;serial=c1dd2d1382ada995;token=OpenBao%20g-k-w;object=bao-root-key;type=public
Cryptoki version 2.40
Manufacturer     SoftHSM
Library          Implementation of PKCS11 (ver 2.6)
go: downloading github.com/openbao/openbao/api/v2 v2.0.1
go: downloading github.com/miekg/pkcs11 v1.1.2-0.20231115102856-9078ad6b9d4b
go: downloading github.com/stretchr/testify v1.8.4
go: downloading github.com/hashicorp/go-uuid v1.0.3
go: downloading google.golang.org/protobuf v1.33.0
go: downloading github.com/pmezard/go-difflib v1.0.0
go: downloading gopkg.in/yaml.v3 v3.0.1
go: downloading github.com/davecgh/go-spew v1.1.1
go: downloading github.com/hashicorp/go-multierror v1.1.1
go: downloading github.com/hashicorp/go-retryablehttp v0.7.7
go: downloading github.com/mitchellh/mapstructure v1.5.0
go: downloading github.com/hashicorp/go-rootcerts v1.0.2
go: downloading github.com/hashicorp/go-secure-stdlib/parseutil v0.1.6
go: downloading golang.org/x/time v0.0.0-20220411224347-583f2d630306
go: downloading golang.org/x/net v0.26.0
go: downloading github.com/go-jose/go-jose/v3 v3.0.1
go: downloading github.com/cenkalti/backoff/v3 v3.0.0
go: downloading github.com/hashicorp/go-cleanhttp v0.5.2
go: downloading github.com/hashicorp/errwrap v1.1.0
go: downloading github.com/hashicorp/hcl v1.0.0
go: downloading github.com/hashicorp/go-sockaddr v1.0.2
go: downloading github.com/hashicorp/go-secure-stdlib/strutil v0.1.2
go: downloading golang.org/x/crypto v0.24.0
go: downloading github.com/ryanuber/go-glob v1.0.0
go: downloading golang.org/x/text v0.16.0
=== RUN   Test_GetOpts
=== PAUSE Test_GetOpts
=== RUN   TestAccPkcs11Wrapper_Lifecycle
    pkcs11_acc_test.go:38: err: failed to pkcs11 EncryptInit: pkcs11: 0x7: CKR_ARGUMENTS_BAD
--- FAIL: TestAccPkcs11Wrapper_Lifecycle (0.01s)
=== CONT  Test_GetOpts
=== RUN   Test_GetOpts/WithKeyId
=== RUN   Test_GetOpts/WithSlot
=== RUN   Test_GetOpts/WithPin
=== RUN   Test_GetOpts/WithLib
=== RUN   Test_GetOpts/WithTokenLabel
=== RUN   Test_GetOpts/WithMechanism
--- PASS: Test_GetOpts (0.00s)
    --- PASS: Test_GetOpts/WithKeyId (0.00s)
    --- PASS: Test_GetOpts/WithSlot (0.00s)
    --- PASS: Test_GetOpts/WithPin (0.00s)
    --- PASS: Test_GetOpts/WithLib (0.00s)
    --- PASS: Test_GetOpts/WithTokenLabel (0.00s)
    --- PASS: Test_GetOpts/WithMechanism (0.00s)
FAIL
FAIL	github.com/openbao/go-kms-wrapping/wrappers/pkcs11/v2	0.010s
FAIL
Error: building at STEP "RUN true && 	dnf update --refresh -y && 	dnf install -y softhsm opensc golang && 	mkdir /softhsm && 	export SOFTHSM2_CONF=/go-kms-wrapping/wrappers/pkcs11/softhsm2.conf && 	softhsm2-util --init-token --slot 0 --label "OpenBao g-k-w" --so-pin 1234 --pin 4321 && 	softhsm2-util --show-slots && 	export BAO_HSM_TOKEN_LABEL="OpenBao g-k-w" && 	export BAO_HSM_PIN=4321 && 	export BAO_HSM_LIB=/usr/lib64/softhsm/libsofthsm.so && 	export BAO_HSM_KEY_LABEL=bao-root-key && 	pkcs11-tool --module "$BAO_HSM_LIB" --token-label "$BAO_HSM_TOKEN_LABEL" --so-pin 1234 --pin 4321 --keypairgen --key-type rsa:4096 --label "$BAO_HSM_KEY_LABEL" && 	pkcs11-tool --module "$BAO_HSM_LIB" --token-label "$BAO_HSM_TOKEN_LABEL" --so-pin 1234 --pin 4321 --show-info && 	cd /go-kms-wrapping/wrappers/pkcs11 && 	export VAULT_ACC=true && 	go test -v ./... && 	true": while running runtime: exit status 1
```

</details>

glatigny commented:
> @glatigny Hmm, what version of SoftHSM are you testing against? I've been pulling `softhsm x86_64 2.6.1-9.fc40` [in the container](https://github.com/openbao/go-kms-wrapping/pull/14#discussion_r1857158113) but both AES and RSA fail:
> 
> AES failure
> RSA failure

SoftHSM2 does not support RSA PKCS OAEP with SHA256 (default hash implemented).
You have to force the usage of SHA1 by adding :
```
export BAO_HSM_RSA_OAEP_HASH="SHA1"
```

Regarding your AES issue, I can reproduce it with your Dockerfile under fedora ; but it's working fine under Ubuntu:
```bash
FROM ubuntu:latest

# Prepare environment
RUN true && \
    apt update && \
    apt install -y ca-certificates softhsm2 opensc golang && \
    mkdir /softhsm && \
    true

COPY . /go-kms-wrapping

# Set up SoftHSM, generate a key, and run tests.
RUN true && \
    export SOFTHSM2_CONF=/go-kms-wrapping/wrappers/pkcs11/softhsm2.conf && \
    softhsm2-util --init-token --slot 0 --label "OpenBao g-k-w" --so-pin 1234 --pin 4321 && \
    softhsm2-util --show-slots && \
    export BAO_HSM_TOKEN_LABEL="OpenBao g-k-w" && \
    export BAO_HSM_PIN=4321 && \
    export BAO_HSM_LIB=/usr/lib/softhsm/libsofthsm2.so && \ 
    export BAO_HSM_KEY_LABEL="bao-root-key" && \
    export BAO_HSM_MECHANISM="CKM_AES_GCM" && \
    pkcs11-tool --module "$BAO_HSM_LIB" --token-label "$BAO_HSM_TOKEN_LABEL" --so-pin 1234 --pin 4321 --keygen --key-type aes:32 --label "$BAO_HSM_KEY_LABEL" && \
    pkcs11-tool --module "$BAO_HSM_LIB" --token-label "$BAO_HSM_TOKEN_LABEL" --so-pin 1234 --pin 4321 --show-info && \
    pkcs11-tool --module "$BAO_HSM_LIB" -O --so-pin 1234 --pin 4321 && \
    cd /go-kms-wrapping/wrappers/pkcs11 && \
    export VAULT_ACC=true && \
    go test -v ./... && \
    true
```

cipherboy commented:
@glatigny Sorry, post-holiday slowness and forgetfulness :-D Yes, forcing SHA-1 worked wonderfully for OAEP. I wonder if PKCS#11 gives us the tools to interrogate the module about parameters it supports, so we can err with a more detailed message? It looks like `CK_MECHANISM_INFO` struct doesn't have enough information in the Flags, so `C_GetMechanismInfo(...)` won't quite do what we want... ...perhaps we just name it `unsafe-sha1` and leave it at that.

---

The reason it fails on Fedora is because of this copy: 

> ```
> memcpy(&aad[0], CK_GCM_PARAMS_PTR(pMechanism->pParameter)->pAAD, CK_GCM_PARAMS_PTR(pMechanism->pParameter)->ulAADLen);
> ```

cipherboy commented on wrappers/pkcs11/pkcs11_client.go:
```diff
@@ -0,0 +1,668 @@
+
+	// Some HSM will ignore the given nonce and generate their own.
+	// That's why we need to free manually the GCM parameters.
+	params := pkcs11.NewGCMParams(nonce, nil, CryptoAesGcmOverhead*8)
```
```suggestion
	params := pkcs11.NewGCMParams(nonce, []byte{}, CryptoAesGcmOverhead*8)
```

cipherboy replied:
@glatigny Edit: this doesn't actually work -- https://github.com/miekg/pkcs11/blob/v1.1.1/params.go#L79-L83 -- the copy is only done when non-zero length. 

I see: 

https://github.com/softhsm/SoftHSMv2/blob/0053e2b09ba2ed60fa06f173eef4ff4978780a0c/src/lib/SoftHSM.cpp#L2353-L2355

Ubuntu must have a newer SoftHSM version; https://github.com/softhsm/SoftHSMv2/commit/bf71ad7ed7864d7a6400309e729dd6b6dfa4107c was fixed several years ago but my source code doesn't have it.

cipherboy replied:
Ahh: https://github.com/softhsm/SoftHSMv2/tags -- the last tag was what Fedora has: https://packages.fedoraproject.org/pkgs/softhsm/softhsm/ but they've not done a release since. Ubuntu must have picked up the patch.

glatigny replied:
I suppose the Debian/Ubuntu release cherry-picked few patches.
See: https://github.com/softhsm/SoftHSMv2/issues/624

glatigny commented on wrappers/pkcs11/pkcs11_client.go:
```diff
@@ -0,0 +1,668 @@
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	params := pkcs11.NewOAEPParams(hash, mgf_hash, pkcs11.CKZ_DATA_SPECIFIED, nil)
```
@cipherboy for RSA OAEP and SoftHSM2, I dug up a bit.

SoftHSM2 fix the usage of SHA1 for RSA OAEP.
First in the function "MechParamCheckRSAPKCSOAEP":
https://github.com/softhsm/SoftHSMv2/blob/2.6.1/src/lib/SoftHSM.cpp#L12379
(and because there is only that hash possible, the parameters are not kept during asym encrypt init)

The second reason is related tothe crypto librairies.
SoftHSM2 can use BOTAN or OpenSSL ; for the OpenSSL implementation it uses 1.x functions such as "RSA_public_encrypt" which is now deprecated in OpenSSL 3.x
CF : https://github.com/softhsm/SoftHSMv2/blob/2.6.1/src/lib/crypto/OSSLRSA.cpp#L1273

RSA_public_encrypt for RSA OEAP has a fixed hash value set to SHA1.
https://docs.openssl.org/master/man3/RSA_public_encrypt/#description
So does SoftHSM2.

In order to be able to change the Hash/MGF Algorithms, it needs to use "EVP_PKEY_encrypt_init_ex" and set the RSA OAEP parameters (padding, oaep hash and mgf1 hash)
https://docs.openssl.org/master/man3/EVP_PKEY_CTX_ctrl/#rsa-parameters

I did not find a PKCS11 way to retrieve if the token supports RSA OAEP with other hash than SHA1.
Regarding how hash are used during RSA OAEP ; it does not seem to be a security issue. Changing the hash reduce the size of the plain text message which can be encrypt (since the hash will be bigger).

cipherboy replied:
I think that claim is supported by https://eprint.iacr.org/2006/223.pdf. Panos is unconcerned about its use here: https://csrc.nist.gov/csrc/media/Projects/crypto-publication-review-project/documents/initial-comments/fips180-4-initial-public-comments-2022.pdf.

So I think, absent contradictory evidence, allowing SHA1 for interoperability purposes is probably fine, even if we'd prefer SHA2+.

cipherboy reviewed (approved):
Thanks @glatigny! Sorry about the terrible delay in getting this merged, appreciate all your updates and investigation :-)
